### ISSUE TITLE
pgbouncer 1.14 can not use auth_query

### ISSUE THREAD

**Issue by Dragonzlx** [pgbouncer/pgbouncer#521]
Hi, i built pgbouner from source code for postgresql 12.3, when i using the auth_query as the interface to get user paasword by running "SELECT usename, passwd FROM pg_shadow WHERE usename=$1" to login into database, but if always failed, could you pls help to explain this and give me a suggestion to fix this problem, thanks.

**1. my pgbouncer version**

[postgres@dragon01 15:48:43 ~]# pgbouncer --version
PgBouncer 1.14.0
libevent 2.0.21-stable
adns: evdns2
tls: OpenSSL 1.0.2k-fips  26 Jan 2017

**2. my pgbouncer configuration and pg_hba.conf settings**

[databases]
db01 = host=192.168.35.210 port=26432 dbname=test
db02 = host=192.168.35.210 port=26432 dbname=repmgr
db03 = host=192.168.35.210 port=26432 dbname=postgres

[users]
postgres = pool_mode=transaction max_user_connections=10
test = pool_mode=session max_user_connections=10
repmgr = pool_mode=statement max_user_connections=10

[pgbouncer]
logfile = /home/postgres/pgbouncer.log
pidfile = /home/postgres/pgbouncer.pid
listen_addr = *
listen_port = 6432
unix_socket_mode = 0777
unix_socket_dir = /tmp
auth_file = /etc/pgbouncer/userlist.txt
auth_type = md5
auth_hba_file = /etc/pgbouncer/bouncer_hba.conf
auth_user = postgres
auth_query = SELECT usename, passwd FROM pg_shadow WHERE usename=$1
admin_users = repmgr
stats_users = repmgr
pool_mode = session
max_client_conn = 1000
max_user_connections = 100
max_db_connections = 100
default_pool_size = 100
autodb_idle_timeout = 3600
application_name_add_host = 1
server_reset_query = RESET ALL; --DEALLOCATE ALL; /* custom */

client_tls_sslmode = require
client_tls_ca_file = /home/postgres/root.crt
client_tls_key_file = /home/postgres/server.key
client_tls_cert_file = /home/postgres/server.crt
client_tls_ciphers = normal

[root@dragon01 ~]# cat /etc/pgbouncer/userlist.txt
"postgres" "md53175bce1d3201d16594cebf9d7eb3f9d"

[root@dragon01 ~]# cat /etc/pgbouncer/bouncer_hba.conf
host  all  postgres  192.168.35.0/24 trust
host  all  repmgr    192.168.35.0/24 md5

[root@dragon01 ~]# cat /appdata/pgsql/pg12/pg_hba.conf
host    repmgr          repmgr          192.168.35.0/24         md5
host    all             postgres        192.168.35.0/24         trust

**3. login failed when using repmgr user**

[postgres@dragon01 16:05:57 ~]# psql -U repmgr -d repmgr -p 6432
Password for user repmgr: 
psql: error: could not connect to server: ERROR:  password authentication failed
ERROR:  SSL required
[postgres@dragon01 16:06:03 ~]# 
[postgres@dragon01 16:06:04 ~]# psql -U postgres -d postgres -p 6432
Password for user postgres: 
psql (12.3)
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

postgres=# \q

**Comment by guybrowne** [comment]
I have noticed a similar issue with auth_user in recent versions of pgbouncer when defining the auth_user parameter under the [pgbouncer] Section.

I have worked around the problem by specifying the auth_user with the database definition in the [databases] Section.

**Comment by Dragonzlx** [comment]
As you said, this can be worked when i set the [database] like this, i will make a test based on the settings on bellow later, thanks a lot.
[databases]
db01 = host=192.168.35.210 port=26432 dbname=test auth_user=postgres
db02 = host=192.168.35.210 port=26432 dbname=repmgr auth_user=postgres
db03 = host=192.168.35.210 port=26432 dbname=postgres

**Comment by Dragonzlx** [comment]
It is still failed to login by defined user, i don't know what happend, could you give me another suggestion to make it worked, thanks a lot.


**1. pgbouncer settings**

[root@dragon01 ~]# cat /etc/pgbouncer.ini
[databases]
db01 = host=192.168.35.210 port=26432 dbname=postgres auth_user=pgbouncer

[users]
test = pool_mode=session max_user_connections=10
repmgr = pool_mode=statement max_user_connections=10
postgres = pool_mode=session max_user_connections=10

[pgbouncer]
logfile = /home/postgres/pgbouncer.log
pidfile = /home/postgres/pgbouncer.pid
listen_addr = *
listen_port = 6432
unix_socket_mode = 0777
unix_socket_dir = /tmp
auth_file = /etc/pgbouncer/userlist.txt
auth_type = hba
auth_hba_file = /etc/pgbouncer/bouncer_hba.conf
auth_user = pgbouncer
auth_query = SELECT * FROM pgbouncer.get_auth($1)
admin_users = pgbouncer
stats_users = pgbouncer
pool_mode = session
max_client_conn = 1000
max_user_connections = 100
max_db_connections = 100
default_pool_size = 100
autodb_idle_timeout = 3600
application_name_add_host = 1
server_reset_query = RESET ALL; --DEALLOCATE ALL; /* custom */
client_tls_sslmode = require
client_tls_ca_file = /home/postgres/root.crt
client_tls_key_file = /home/postgres/server.key
client_tls_cert_file = /home/postgres/server.crt
client_tls_ciphers = normal


[root@dragon01 ~]# cat /etc/pgbouncer/userlist.txt 
"pgbouncer" "md5003c7242487c89ca07282f6e048a4821"


[root@dragon01 ~]# cat /etc/pgbouncer/bouncer_hba.conf 
host  all  repmgr    192.168.35.0/24 md5
host  all  pgbouncer 192.168.35.0/24 trust
host  all  test      192.168.35.0/24 md5
host  all  postgres  192.168.35.0/24 md5

**2. postgresql pg_hba.conf**

host    postgres        repmgr          192.168.35.0/24         md5
host    all             pgbouncer       192.168.35.0/24         trust
host    all             test            192.168.35.0/24         md5

3.  i can not login by user postgres and ok to login by pgbouncer

[postgres@dragon01 17:46:01 /appdata/pgsql/pg12]# psql -U pgbouncer -d db01 -p 6432      
psql (12.3)
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

db01=> \q

[postgres@dragon01 17:46:14 /appdata/pgsql/pg12]# psql -U postgres -d db01 -p 6432 
Password for user postgres: 
psql: error: could not connect to server: ERROR:  password authentication failed
ERROR:  SSL required

[postgres@dragon01 17:46:24 /appdata/pgsql/pg12]# psql -U postgres -p 26432       
Password for user postgres: 
psql (12.3)
Type "help" for help.

postgres=# \q


**4. pgbouncer have the privileges to execute on  pgbouncer.get_auth**

[postgres@dragon01 17:50:15 /appdata/pgsql/pg12]# psql -U pgbouncer -d postgres -p 26432
psql (12.3)
Type "help" for help.

postgres=> SELECT * FROM pgbouncer.get_auth('postgres');
NOTICE:  PgBouncer auth request: postgres
 username |              password               
----------+-------------------------------------
 postgres | md53175bce1d3201d16594cebf9d7eb3f9d
(1 row)

**Comment by guybrowne** [comment]
Something I have just discovered but worth trying in your setup, mind you I don't use SSL in my configuration.

I have found that any user defined in the [users] section only works if that user also exists in the authentication file (userlist.txt in this case).  The users I have defined are postgres, pgbouncer and dbmonitor which we use for monitoring.  So this has always worked in my case.  But if I add a user to the [users] section which does not exist in the authentication file (userlist.txt) I get password authentication failed... even though this user has connect access to all databases.

I even set the auth_user=postgres so has superuser access and directly hitting pg_shadow (no function) and get the same issue.

I can confirm this on pgbouncer 1.12 hitting postgresql 11.7 along with pgbouncer 1.14 hitting postgresql 12.4

Also see issue #502 which was the only other time I have tried using a defined user configuration and that user was not in the authentication file.  That pgbouncer configuration was not using pgbouncer.get_auth function and worked fine with the auth_user just defined in the [pgbouncer] section.
